src/main/java/org/myrobotlab/service/NeoPixel.java

```diff
@@ -26,501 +26,829 @@
-    setAnimation(animationStringToInt(animation), Integer.parseInt(red), Integer.parseInt(green), Integer.parseInt(blue), Integer.parseInt(speed));
-  }
-
-  int animationStringToInt(String animation) {
-    switch (animation) {
-      case "No animation":
-        return NEOPIXEL_ANIMATION_STOP;
-      case "Stop":
-        return NEOPIXEL_ANIMATION_STOP;
-      case "Color Wipe":
-        return NEOPIXEL_ANIMATION_COLOR_WIPE;
-      case "Larson Scanner":
-        return NEOPIXEL_ANIMATION_LARSON_SCANNER;
-      case "Theater Chase":
-        return NEOPIXEL_ANIMATION_THEATER_CHASE;
-      case "Theater Chase Rainbow":
-        return NEOPIXEL_ANIMATION_THEATER_CHASE_RAINBOW;
-      case "Rainbow":
-        return NEOPIXEL_ANIMATION_RAINBOW;
-      case "Rainbow Cycle":
-        return NEOPIXEL_ANIMATION_RAINBOW_CYCLE;
-      case "Flash Random":
-        return NEOPIXEL_ANIMATION_FLASH_RANDOM;
-      case "Ironman":
-        return NEOPIXEL_ANIMATION_IRONMAN;
-      default:
-        log.error("Unknow Animation type {}", animation);
-        return NEOPIXEL_ANIMATION_STOP;
+  public void setMatrix(int address, int red, int green, int blue) {
+    setMatrix(address, red, green, blue, 0);
+  }
+
+  public void setMatrix(int address, int red, int green, int blue, int white) {
+    Pixel p = getPixelSet().pixels.get(address);
+    p.red = red;
+    p.green = green;
+    p.blue = blue;
+    p.white = white;
+  }
+
+  public void setPin(int pin) {
+    this.pin = pin;
+    broadcastState();
+  }
+
+  /**
+   * basic setting of a pixel
+   */
+  public void setPixel(int address, int red, int green, int blue) {
+    setPixel(currentMatrix, currentSequence, address, red, green, blue, 0, 0);
+  }
+
+  public void setPixel(int address, int red, int green, int blue, int white) {
+    setPixel(currentMatrix, currentSequence, address, red, green, blue, white, 0);
+  }
+
+  /**
+   * setPixel of maximum complexity
+   * 
+   * @param matrixName
+   * @param pixelSetIndex
+   * @param address
+   * @param red
+   * @param green
+   * @param blue
+   * @param white
+   * @param delayMs
+   */
+  public void setPixel(String matrixName, Integer pixelSetIndex, int address, int red, int green, int blue, int white, Integer delayMs) {
+
+    // get and update memory cache
+    PixelSet ps = getPixelSet(matrixName, pixelSetIndex);
+
+    ps.delayMs = delayMs;
+
+    // NeoPixelController c = (NeoPixelController)
+    // Runtime.getService(controller);
+    ServiceInterface sc = Runtime.getService(controller);
+    if (sc == null) {
+      error("controler %s not valid", controller);
+      return;
     }
 
+    // update pixel in sequence in matrix
+    Pixel pixel = new Pixel(address, red, green, blue, white);
+
+    // update memory
+    ps.pixels.set(address, pixel);
+
+    // write immediately
+    NeoPixelController np2 = (NeoPixelController) Runtime.getService(controller);

```

Comment:
This is silly, why do we need to get a handle to the controller each time?

Reply:
one step closer to true pub/sub